### ISSUE TITLE
keylog_callback not called for TLS 1.3 re-key

### ISSUE THREAD

**Issue by meins111** [openssl/openssl#19226]
Hi,

I noticed that the keylog_callback (which you can set via calling [set_keylog_callback](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_keylog_callback.html) is not called by the key update handler.
This means that the initial traffic secret(s) can be logged but any subsequent traffic secrets will not.

The man page tells us the intended behavior:
"This callback is called whenever TLS key material is generated or received".

From a brief glance at the src, it may be as simple as adding another call to `ssl_log_secret` within the `tls13_update_key` method (in tls13:enc.c) but I do not know the code well enough to judge this properly.

**Comment by hlandau** [comment]
This could be considered a bug, but I'm not sure it necessarily is. In TLS 1.3, the updated traffic secret is computed from the previous traffic secret. The purpose of key update in TLS 1.3 is to avoid exceeding AEAD usage limits for a given key; the process does not inject any more entropy into a connection's cryptogrpahic materials. Thus the secrets currently logged are adequate to determine the keys after a key update: https://www.rfc-editor.org/rfc/rfc8446#section-7.2

**Comment by meins111** [comment]
Yes, I looked it up and even have written a workaround using the message callback to notice key update message to then be able to compute the new traffic secret myself but that's really annoying and I ran into even more trouble from there (as I have to be very careful to not perform "too many" key updates, if multiple sources detect the key update etc.

Anyway, this probably is borderline feature request / bug, although the man page entry seems pretty clear cut on when it is called...

Would you be interested if I were to give it a shot for a PR/patch?

**Comment by mattcaswell** [comment]
I'd personally put it on the "bug" side of the line. I'd be happy to see a PR for it. Adding the missing `ssl_log_secret` call seems a reasonable approach.

**Comment by hlandau** [comment]
@mattcaswell Indeed, I could see explicitly logging the subsequent traffic keys being useful if you start capturing a stream halfway through. Though this is not too useful in ordinary TLS since the nonce counter value will be unknown. For DTLS I imagine it would be useful though, so for consistency we may as well do it for both.

**Comment by mattcaswell** [comment]
>  I could see explicitly logging the subsequent traffic keys being useful if you start capturing a stream halfway through.

Even if you are logging from the beginning it is still useful for TLS. Otherwise any consumer of the log will have to be watching out for key update messages and implement all the logic to compute the new keys. I think that was the original intent.

**Comment by meins111** [comment]
And just happens to be exactly what I was trying to do and why I ran into this. 

And as for the start_anywhere TLS observer, you can always try to guess the sequence number if you have the proper keys and if you get a notify upon a re-key event you can immediately "join" as the sequence number will be reset to 0 then.

Formality wise: PR to master or Openssl3.0 branch? Will be my first one here, so not 100% on the repo structure...

**Comment by hlandau** [comment]
If you make a PR, you can just make it against master. If we decide it's a bug, we'll also merge it to 3.0 if it merges cleanly, or failing that ask you to make another PR for 3.0. If we decide it's a feature, it will go into master only.

**Comment by meins111** [comment]
It seems that the patch was as easy as I though.
I have a local build with passing (openssl) tests and I do get the updates in my own project relying on the callback being triggered upon a key update - yay.

I also tried (but failed) to create a new branch (and subsequent PR) in this repo. I do however, get a 403 - Permission Denied error trying to push my local branch to https://github.com/openssl/openssl.git/
Do I need some special permission from your side to be able to do so? Or is it a problem on my side?
I can push to my own repo with no issue however, so I felt this was maybe something to do with the openssl repo config?

[0001-ssl_log_secret-call-added-within-tls13_update_key.zip](https://github.com/openssl/openssl/files/9596150/0001-ssl_log_secret-call-added-within-tls13_update_key.zip)

**Comment by meins111** [comment]
https://github.com/openssl/openssl/pull/19241
